ignition-server/controllers/machineconfigserver_ignitionprovider.go

```diff
@@ -0,0 +1,338 @@
+)
+
+var _ IgnitionProvider = (*MCSIgnitionProvider)(nil)
+
+// MCSIgnitionProvider is an IgnitionProvider that uses
+// MachineConfigServer pods to build ignition payload contents.
+type MCSIgnitionProvider struct {
+	Client          client.Client
+	ReleaseProvider releaseinfo.Provider
+	Namespace       string
+}
+
+func (p *MCSIgnitionProvider) GetPayload(ctx context.Context, tokenSecret *corev1.Secret) (payload []byte, err error) {
+	// TODO(alberto): If the MCS supports binding address
+	// https://github.com/openshift/machine-config-operator/pull/2630/files
+	// we could bind it to localhost and get the payload by execing
+	// https://zhimin-wen.medium.com/programing-exec-into-a-pod-5f2a70bd93bb
+	// Otherwise with current approach the mcs pod is temporary exposed in the pod network every time a payload is generated.
+	releaseImage := string(tokenSecret.Data[TokenSecretReleaseKey])
+	img, err := p.ReleaseProvider.Lookup(ctx, releaseImage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to look up release image metadata: %w", err)
+	}
+
+	mcsResourceName := tokenSecret.GetName()
+	mcsServiceAccount := MachineConfigServerServiceAccount(p.Namespace, mcsResourceName)
+	mcsRoleBinding := MachineConfigServerRoleBinding(p.Namespace, mcsServiceAccount)
+	mcsPod := mcsPod(p.Namespace, mcsResourceName, img, mcsServiceAccount)
+	// Launch the pod and ensure we clean up regardless of outcome
+	if err := p.Client.Create(ctx, mcsPod); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server pod: %w", err)
+	}
+
+	if err := p.Client.Create(ctx, mcsServiceAccount); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server Service Account: %w", err)
+	}
+
+	if err := p.Client.Create(ctx, mcsRoleBinding); err != nil {
+		return nil, fmt.Errorf("failed to create machine config server Role Binding: %w", err)
+	}
+	defer func() {
+		var deleteErrors []error
+		if err := p.Client.Delete(ctx, mcsServiceAccount); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		if err := p.Client.Delete(ctx, mcsRoleBinding); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		if err := p.Client.Delete(ctx, mcsPod); err != nil && errors.IsNotFound(err) {
+			deleteErrors = append(deleteErrors, fmt.Errorf("failed to delete machine config server pod: %w", err))
+		}
+		// We return this in the named returned variable.
+		err = utilerrors.NewAggregate(deleteErrors)
+	}()
+
+	// If the machine config server is not ready we return and wait for an update event to reconcile.
+	mcsReady := false
+	for _, cond := range mcsPod.Status.Conditions {
+		if cond.Type == corev1.ContainersReady && cond.Status == corev1.ConditionTrue {
+			mcsReady = true
+		}
+	}
+	if mcsPod.Status.PodIP == "" || !mcsReady {
+		return nil, nil
+	}
```

Comment:
The interface for this function was designed as a synchronous library call — won't you need to wait here?

Reply:
I see no need to wait since the controller will reconcile on changes to the pod, I think is a fair assumption for the provider implementation since we own both.

Reply:
I think the lifecycle of everything needed to get the payload (the MCS pod, its SA, rolebinding etc) should be scoped entirely to the life of this function call. Otherwise this isn't a library call, it's a full on controller doing reconciliation, which would be a different design entirely IMO, and I don't know what the semantics of the function would be... if the pod isn't ready, what does the caller get?

If I call GetPayload() I expect either the result or a failure to get the result, with no side effects (e.g. pods left over etc). The fact we're lifecycling a pod behind the call is an implementation detail, and the function was intended to be synchronous and all-or-nothing.